ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Block.hs

```diff
@@ -97,14 +98,21 @@ import           Ouroboros.Consensus.Shelley.Ledger (ShelleyBlock)
 -- TODO: parameterise ByronBlock over crypto too
 type CardanoEras c =
   '[ ByronBlock
-   , ShelleyBlock (ShelleyEra c)
-   , ShelleyBlock (AllegraEra c)
-   , ShelleyBlock (MaryEra c)
-   , ShelleyBlock (AlonzoEra c)
+   , ShelleyBlock (TPraos c) (ShelleyEra c)
+   , ShelleyBlock (TPraos c) (AllegraEra c)
+   , ShelleyBlock (TPraos c) (MaryEra c)
+   , ShelleyBlock (TPraos c) (AlonzoEra c)
    ]
 
 -- | The Shelley-based eras in the Cardano chain
-type ShelleyBasedEras c = '[ShelleyEra c, AllegraEra c, MaryEra c, AlonzoEra c]
+type ShelleyErasAndProtos c =
+  [ '(TPraos c, ShelleyEra c)
```

Comment:
Here's a thought. If we used

```
type family ShelleyBlockProto blk where ShelleyBlockProto (ShelleyBlock proto era) = proto
type family ShelleyBlockEra   blk where ShelleyBlockEra   (ShelleyBlock proto era) = era
```

instead of `Fst` and `Snd`, then we could use `ShelleyBlock` instead of `(,)`.

Benefits:

- Most (?) of the infra you added wouldn't be necessary.
- We wouldn't need a separate `ShelleyErasAndProtos` list, we could just have `ShelleyBasedEras` as per usual.

Reply:
Urgh, that sounds disturbingly plausible

Reply:
Unfortunately this doesn't actually work terribly well for us (or, at least, in a simple implementation of this). The new machinery is mostly to deal with the fact that `ShelleyBlock` takes two arguments, so is still needed. Once we need that, the tupled approach then seems the simplest to apply things to that 2-argument functor.

Reply:
Yeah, sounds plausible. If the generic stuff we specialized to do things like `f (ShelleyBlockProto x) (ShelleyBlockEra x)` then its kinds might be a bit simpler. But we'd still need the uncurrying, basically. It'd just be more obvious that all this machinery is for `ShelleyBlock`.

Reply:
I've now removed this by another means.